File: src/main/java/edu/kit/iti/formal/automation/testtables/model/TableNode.java
package edu.kit.iti.formal.automation.testtables.model;

import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.util.LinkedList;
import java.util.List;

/**
 * @author Alexander Weigl
 * @version 1 (01.02.18)
 */
@Data
@RequiredArgsConstructor
public abstract class TableNode {
    protected final int id;
    protected Duration duration = new Duration(1, 1);

    public abstract boolean isLeaf();

    public abstract List<TableNode> getChildren();

    public abstract int count();

    public abstract List<State> flat();

    public abstract int depth();

    public abstract List<State.AutomatonState> getAutomataStates();
}

File: geteta/src/main/kotlin/edu/kit/iti/formal/automation/testtables/model/TableNode.kt
/*
 * geteta
 *
 * Copyright (C) 2016-2018 -- Alexander Weigl <weigl@kit.edu>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 */
package edu.kit.iti.formal.automation.testtables.model

import lombok.Data
import lombok.RequiredArgsConstructor

/**
 * @author Alexander Weigl
 * @version 1 (01.02.18)
 */
abstract class TableNode(val id: String) {
    var duration = Duration(1, 1)

    abstract val isLeaf: Boolean
    abstract val children: List<TableNode>
    abstract val automataStates: List<State.AutomatonState>

    abstract fun count(): Int
    abstract fun flat(): List<State>
    abstract fun depth(): Int

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as TableNode
        if (id != other.id) return false
        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}